### DISCUSSION TITLE
Clarification on type annotations in encoded output

### DISCUSSION THREAD

**Discussion by benjreinhart** [kdl-org/kdl#232] [Q&A] [answered]
Hello all,

Regarding type annotations, the [spec](https://github.com/kdl-org/kdl/blob/main/SPEC.md#type-annotation) states:

> KDL does not specify any restrictions on what implementations might do with these annotations. They are free to ignore them, or use them to make decisions about how to interpret a value.

Given that, I would not necessarily expect encoded output to have to retain type annotations for values. However, it looks like there are a number of test cases whose encoded output does contain the type annotation.

My question is: are implementations required to retain type annotation information for encoding? Or are the tests perhaps outdated?

FWIW, the rust and JS implementations do not seem to pass these tests, but maybe they just haven't implemented this behavior yet.

Many examples in [this commit](https://github.com/kdl-org/kdl/commit/f59ca2476aa8e79e6bf78c7ec18db20f26054e7b). One example [here](https://github.com/kdl-org/kdl/blob/main/tests/test_cases/expected_kdl/arg_false_type.kdl):

input:

```kdl
node (type)false
```

output:

```kdl
node (type)false
```

**Comment by tabatkins** [comment, marked as answer]
Yes, type annotations are part of the data model. Impls just don't have to *do* anything with them; they can treat the node/value exactly like normal otherwise.

**Reply by benjreinhart** [reply]
Do any implementations expose this behavior yet that I could take a look at? It doesn't look like the ones that are a part of this org do. 

Asking because it would seem that implementations would need to expose the annotations anywhere there is a value, so instead of returning a list of values it'd be a list of `(type, value)`, right? For example, the rust behavior is as follows:

Input:

```
node (type)false
```

Parse result:

```rust
[
   KdlNode { name: "node", values: [Boolean(false)], properties: {}, children: [] }
]
```

The type annotation has been lost, so encoding that would yield `node false`, not `node (type)false`. In order to pass that test, implementations would need to expose the type annotation back to the user even if it does nothing with it, correct?

**Comment by tabatkins** [comment]
Yes, you need to preserve the tag with a custom class in this case. <https://github.com/tabatkins/kdlpy> does this, for instance - by default it'll parse *untagged* values into native Python types, but tagged values become lib-specific types.

**Reply by benjreinhart** [reply]
Great, thank you!

**Comment by hkolbeck** [comment]
kdl4j retains them and round trips them, but currently they aren't used internally.
